### PR TITLE
[EOS-13675]: Rethink storage_type value detection method

### PR DESCRIPTION
**Pull request by dmitrydb**
Change logic for storage_type detection. Use pillar based approach the first(values fills from user's **config.ini**). If it fails, use smart detection approach for that. Update **storage_enclosure/type** pillar value if it was missed originally.

### CODE REVIEW

File: api/python/provisioner/commands/get_setup_info.py [hunk 1]
```diff
@@ -196,36 +196,27 @@ def _get_storage_type_pillar_based():
...
-                controller_type
-            )
+            (storage_enclosure_type,)
         )
 
         pillar = pillar.get(local_minion_id())  # type: dict
 
-        for key in (storage_enclosure_type, controller_type):
-            if not pillar[key] or pillar[key] is values.MISSED:
+        for key in (storage_enclosure_type,):
```
Comment (andkononykhin2) [inline api/python/provisioner/commands/get_setup_info.py:206]: I believe loop for is not needed

File: api/python/provisioner/commands/get_setup_info.py [hunk 2]
```diff
@@ -278,14 +269,92 @@ def _get_storage_type(self):
...
+            _update_storage_type_pillar(_res.get(config.STORAGE_TYPE))
+            return _res
+
+        logger.debug("Storage smart detection function returned storage_type "
+                     f"'{_res[config.STORAGE_TYPE]}'")
+
+        # try to detect storage_type using the original method via controller
+        # type pillar value
+        res = dict()
+        storage_enclosure_path = KeyPath('storage_enclosure')
```
Comment (andkononykhin2) [inline api/python/provisioner/commands/get_setup_info.py:331]: Please move that to a separate method. Update fun might be separated as a class nethod as well. So here you would have only upper level logic. More modular and clear.

**Review by andkononykhin2 — CHANGES_REQUESTED** [review]

File: api/python/provisioner/commands/get_setup_info.py [hunk 3]
```diff
@@ -278,14 +269,92 @@ def _get_storage_type(self):
...
+            _storage_enclosure_type = PillarKey(
+                                        _storage_enclosure_path / 'type')
+            pillar_updater = PillarUpdater(config.ALL_MINIONS)
+            pillar_updater.update((_storage_enclosure_type, storage_type))
+
+        # Get storage type from pillar values
+        try:
+            return self._get_storage_type_pillar_based()
+        except BadPillarDataError:
+            logger.debug("Pillar based approach for storage_type detection "
```
[the remark this answers is not in the document]
Comment (dmitrydb) [reply api/python/provisioner/commands/get_setup_info.py:315]: Improve that

File: api/python/provisioner/commands/get_setup_info.py [hunk 4]
```diff
@@ -278,14 +269,92 @@ def _get_storage_type(self):
...
+
+            :param storage_type: new storage_type value for pillar
+                                storage_enclosure/type
+            :return:
+            """
+            _storage_enclosure_path = KeyPath('storage_enclosure')
+            _storage_enclosure_type = PillarKey(
+                                        _storage_enclosure_path / 'type')
+            pillar_updater = PillarUpdater(config.ALL_MINIONS)
+            pillar_updater.update((_storage_enclosure_type, storage_type))
```
[the remark this answers is not in the document]
Comment (dmitrydb) [reply api/python/provisioner/commands/get_setup_info.py:309]: Done

File: api/python/provisioner/commands/get_setup_info.py [hunk 5]
```diff
@@ -278,14 +269,92 @@ def _get_storage_type(self):
...
+
+        if pillar[controller_type] == config.ControllerTypes.GALLIUM.value:
+            res[config.STORAGE_TYPE] = config.StorageType.ENCLOSURE.value
+        elif pillar[controller_type] == config.ControllerTypes.INDIUM.value:
+            res[config.STORAGE_TYPE] = config.StorageType.PODS.value
+        else:
+            res[config.STORAGE_TYPE] = config.StorageType.OTHER.value
+        # TODO: what controller type for JBOD and EBOD?
+
+        if res.get(config.STORAGE_TYPE) != config.StorageType.OTHER.value:
```
[the remark this answers is not in the document]
Comment (dmitrydb) [reply api/python/provisioner/commands/get_setup_info.py]: It is no more needed as per discussion with @ipoddubnyy we need to remove any mention of 5u84/PODS and so on. The same for controller type names

File: api/python/provisioner/commands/get_setup_info.py (hunk 1 above)
Comment (dmitrydb) [reply api/python/provisioner/commands/get_setup_info.py]: Fixed

File: api/python/provisioner/commands/get_setup_info.py (hunk 2 above)
Comment (dmitrydb) [reply api/python/provisioner/commands/get_setup_info.py]: Fixed.

File: api/python/provisioner/commands/get_setup_info.py [hunk 6]
```diff
@@ -278,14 +269,92 @@ def _get_storage_type(self):
...
+                         "failed")
+
+        # detect storage type using info about system configuration
+        _res = self._detect_storage_type()
+
+        if _res.get(config.STORAGE_TYPE) != config.StorageType.OTHER.value:
+            _update_storage_type_pillar(_res.get(config.STORAGE_TYPE))
+            return _res
+
+        logger.debug("Storage smart detection function returned storage_type "
```
[the remark this answers is not in the document]
Comment (dmitrydb) [reply api/python/provisioner/commands/get_setup_info.py]: It is no more needed as per discussion with @ipoddubnyy we need to remove any mention of 5u84/PODS and so on. The same for controller type names

**Review by ypise — APPROVED** [review]

**Comment by ypise** [comment]
@dmitrydb 
I've changed the base to pre-cortx-1.0. This would be same as cortx-1.0 with one additional commit for M5.
Do rebase to ensure this branch is aligned.

**Comment by dmitrydb** [comment]
@ypise Merged `pre-cortx-1.0` target branch to current one.

File: pillar/components/samples/generic_config.ini [hunk 7]
```diff
@@ -8,6 +8,7 @@ controller.primary_mc.ip=10.0.0.1
 controller.secondary_mc.ip=10.0.0.2
 controller.user=manage
 controller.secret='!manage'
+controller.type=
```
Comment (andkononykhin2) [inline pillar/components/samples/generic_config.ini:11]: @dmitrydb i guess that field is not supported currently by ConfigureSetup command. It should pass the static validation and be set to o pillar then.

**Review by andkononykhin2 — CHANGES_REQUESTED** [review]

File: pillar/components/samples/generic_config.ini (hunk 7 above)
Comment (dmitrydb) [reply pillar/components/samples/generic_config.ini:11]: I checked configure_setup command with updated by Yash **generic_config.in**i. This template successfully passes validation of **StorageEnclosureParamsValidation**. If you want I can add it to params validation logic and set it as mandatory param for **storage_enclosure** section

**Review by ypise — APPROVED** [review]

**Pull request merged**